libraries/AP_Airspeed/AP_Airspeed.cpp

```diff
@@ -399,15 +427,122 @@ void AP_Airspeed::read(void)
+        }
+    }
+    
+    // calculating difference
+    for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+        ekf_sensor_airspeed_diff[i] = fabsf(state[i].airspeed - airspeed_ekf);
+    }
+    
+    if (healthy_sens_num > 0)
+    {
+        airspeed_avg /= healthy_sens_num;
+        
+        const float airspeed_avg_limit = airspeed_ekf * AIRSPEED_BLEND_DIFF_LIMIT;
+
+        // look for healthy sensors that might be failing
+
+        // Idea is to find the sensor which differs much both from other sensors and from ekf value
+        // If there are two sensors then the first check will be true for both, however
+        // the second check will be true only for the sensor which really failed
+        for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+            if (state[i].blend_healthy &&
+                    ekf_sensor_airspeed_diff[i] > airspeed_avg_limit && state[i].airspeed > AIRSPEED_BLEND_MIN_SPEED) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Airspeed[%u] sensor blending unhealthy", i);
+                state[i].blend_healthy = false;
+            }
+        }
+
+        // TODO: Check what is really needed and where is used. Maybe something is missed.
+        blend_state.airspeed = 0.0;
+        blend_state.raw_airspeed = 0.0;
+        blend_state.temperature = 0.0;
+        blend_state.last_pressure = 0.0;
+        blend_state.corrected_pressure = 0.0;
+        blend_state.EAS2TAS = 0.0;
+        blend_state.ratio = 0.0;
+        blend_state.offset = 0.0;
+
+        healthy_sens_num = 0;
+
+        for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+            if (!state[i].blend_healthy) {
+                continue;
+            }
+            
+            blend_state.airspeed += state[i].airspeed;
+            blend_state.raw_airspeed += state[i].raw_airspeed;
+
+            // TODO: Check for get_temperature result and correction to number of sensors
+            float temperature;
+            get_temperature(i, temperature);
+            blend_state.temperature += temperature;
+
+            blend_state.last_pressure += state[i].last_pressure;
+            blend_state.corrected_pressure += state[i].corrected_pressure;
+            blend_state.EAS2TAS += state[i].EAS2TAS;
+            
+            blend_state.ratio += param[i].ratio;
+            blend_state.offset += param[i].offset;
+
+            healthy_sens_num++;
+        }
+        
+        if (healthy_sens_num > 0) {
+            float div2mul_factor = 1.0 / healthy_sens_num;
+            blend_state.airspeed *= div2mul_factor;
+            blend_state.raw_airspeed *= div2mul_factor;
+            blend_state.temperature *= div2mul_factor;
+            blend_state.last_pressure *= div2mul_factor;
+            blend_state.corrected_pressure *= div2mul_factor;
+            blend_state.EAS2TAS *= div2mul_factor;
+            blend_state.ratio *= div2mul_factor;
+            blend_state.offset *= div2mul_factor;
```

Comment:
I'll let this pass on this one because it's not a problem but if I was to write this I would not pass blend_state through this rollercoaster. In a threaded world if something was looking at that blend_state at the wrong time it would catch the variable in a crazy temporary state because it's being written to (AIRSPEED_MAX_SENSORS+2) times. A better solution is:
```
blend_state_t blend_state_temp {}; // sets everything to zero
// loop to increment blend_state_temp.*
// divide down blend_state_temp.* to get the avg
memcpy(blend_state, blend_state_temp, sizeof(blend_state)); // blend_state is written to once

```

Reply:
Well, in a threaded world I would not consider memcpy to be atomic either. So real solution would be mutexes. Should we go for it?

Reply:
Nah. This library doesn't suffer from threading problems like this so I wouldn't worry about it. And yes, memcpy isn't thread-proof either but it's close. I just hate seeing values go on a rollercoaster like that.

Reply:
That depends on OS, considering a few already in AP.

Reply:
Done